base/dates/rounding.jl

```diff
@@ -77,6 +80,35 @@ function Base.floor(dt::DateTime, p::TimePeriod)
 end
 
 """
+    floor(x::Period, precision::T) where T <: Union{TimePeriod, Week, Day} -> T
+
+Rounds `x` down to the nearest multiple of `precision`. If `x` and `precision` are different
+subtypes of `Period`, the return value will have the same type as `precision`.
+
+For convenience, `precision` may be a type instead of a value: `floor(x, Dates.Hour)` is a
+shortcut for `floor(x, Dates.Hour(1))`.
+
+```jldoctest
+julia> floor(Dates.Day(16), Dates.Week)
+2 weeks
+
+julia> floor(Dates.Minute(44), Dates.Minute(15))
+30 minutes
+
+julia> floor(Dates.Hour(36), Dates.Day)
+1 day
+```
+
+Rounding to a `precision` of `Month`s or `Year`s is not supported, as these `Period`s are of
+inconsistent length.
+"""
+function Base.floor(x::ConvertiblePeriod, precision::T) where T <: ConvertiblePeriod
+    value(precision) < 1 && throw(DomainError(precision))
+    x, precision = promote(x, precision)
```

Comment:
Should use a different variable name for the assignment to ensure type stability